### DISCUSSION TITLE
Introduce foreign keys in database schema

### DISCUSSION THREAD

**Discussion by flavius** [PrestaShop/PrestaShop#31100] [Ideas]
Dear PrestaTeam

Are there any plans or thoughts on referential integrity, at least for new installations?

> Reasons to use Foreign Keys:
> 
> you won't get Orphaned Rows (real benefit for PS!)
> you can get nice "on delete cascade" behavior, automatically cleaning up tables (real benefit for PS!)
> knowing about the relationships between tables in the database helps the Optimizer plan your queries for most efficient execution, since it is able to get better estimates on join cardinality. (real benefit for PS!!)
> FKs give a pretty big hint on what statistics are most important to collect on the database, which in turn leads to better performance (real benefit for PS!!)
> they enable all kinds of auto-generated support -- ORMs can generate themselves, visualization tools will be able to create nice schema layouts for you, etc (will help PS developers)
> someone new to the project will get into the flow of things faster since otherwise implicit relationships are explicitly documented (will help PS developers)

**Comment by khouloudbelguith** [comment]
Ping @PrestaShop/prestashop-core-developers what do you think?

Thanks!

**Comment by flavius** [comment]
Ping @PrestaShop/prestashop-core-developers ?

Ping @khouloudbelguith ?

**Comment by khouloudbelguith** [comment]
@PrestaShop/prestashop-core-developers friendly reminder :)

**Comment by PierreRambaud** [comment]
Hi and thanks for your suggestion.
We currently plan nothing about adding it. But I'm in favor of adding referential integrity with this kind of database (mysql, pgsql, etc).
ping @eternoendless @matks @jolelievre

**Comment by jolelievre** [comment]
I think it could be great as well, although I don't know why we have no foreign keys for now
Maybe is it because of the modules that can change the database and would me more difficult to deal with with foreign keys? 🤔
Anyway if in some next major version we decide to use Doctrine for all our objects and get rid of the legacy ObjectModel maybe we could introduce FK at this moment, with some nice cascade behaviour

**Comment by Progi1984** [comment]
> Maybe is it because of the modules that can change the database and would me more difficult to deal with with foreign keys? thinking

But it's not a good practice to alter a table for a module. The good practice should be to create new tables with foreign keys.

> Anyway if in some next major version we decide to use Doctrine for all our objects and get rid of the legacy ObjectModel maybe we could introduce FK at this moment, with some nice cascade behaviour

:+1:

**Comment by PierreRambaud** [comment]
@jolelievre It was because of MyISAM which do not support relation keys.

**Comment by flavius** [comment]
@PrestaShop/prestashop-core-developers Thanks for the nice history recap. So how would you like to move forward?

ping @khouloudbelguith

**Comment by eternoendless** [comment]
This is definitely something we should try.

Some history here: https://github.com/PrestaShop/docs/issues/639#issuecomment-661782014

**Comment by nelero** [comment]
Hi, it would be a great improvement for the sanity of the database and particular behaviour with modules or even programmers databases removes which lets orphans in some tables and then create problems.

**Comment by galerienv** [comment]
Hi, why don't implement foreign key without constraint in first step.

Sample : ALTER TABLE PREFIX_address_format ADD CONSTRAINT `FK_Adress_format_Country` FOREIGN KEY (`id_country`) REFERENCES `PREFIX_country` (`id_country`) ON DELETE NO ACTION ON UPDATE NO ACTION;

Align all database field if you have differents size, type in an Second step  

Third step can be implement constraints table by table or more during prestashop patching

If you await to have a clean and aligned field in database to implement foreign key, in 10 years we are at same point.

**Reply by eternoendless** [reply]
I'm not against it!

**Reply by Hlavtox** [reply]
Ping @kpodemski @eternoendless we could discuss it someday 😎

**Comment by prestaalba** [comment]
I made some real test about this topic some time ago. The practical results is that queries are NOT executed faster when tables has foreign keys. So the benefits about DB structure and organization are there, but query performance is not benefited. In most cases, DB performance is the bottleneck for large stores, so these changes should be closely monitored.

**Reply by galerienv** [reply]
The main objective of this proposal is:
1 - have a clean database
2 - have a consistent database
3 - Simply delete fields in the database. who said i was thinking about guest table,connection without orphan
4 - Avoid field duplicates in different tables

Performance was never the main focus

**Comment by mikeartix** [comment]
If this ever gets implemented, please be aware that the engine of the tables might still be `MyISAM` even if the `_MYSQL_ENGINE_` value is set to `InnoDB`. I have faced this issue before with a client, I suspect it's a remnant from upgrading from older versions of PS. Just something to keep in mind.

**Reply by eternoendless** [reply]
Indeed, this could perhaps be fixed by the upgrade module.

**Reply by davidglezz** [reply]
Its super easy! `ALTER TABLE database_name.table_name ENGINE=InnoDB;`

**Reply by eternoendless** [reply]
The case where there are key ref integrity problems will have to be handled though

**Comment by NikolaSivkov** [comment]
Referential integrity is more important than database performance. even if you have million products, all the data can fit into memory of any modern server. Having duplicate & orphaned records can mess your analytics & reporting. 
If you want more performance, cache things! Most of the caching modules do decent enough job to be able to handle tens of thousands of orders a day. I personally manage few stores that have 2-3milion page views a month on 16gb Ram server that hosts both the php & mysql instance. 
I'm baffled how such a popular product as such poor database design choices in 2023.

**Comment by nelero** [comment]
So if we are all agree with doing it, when do we start it to improve things
?


Le dim. 6 août 2023, 17:26, Nikola Sivkov ***@***.***> a
écrit :

**Comment by eternoendless** [comment]
Consider opening a Pull Request with the change! The only thing is that the upgrade module will have to be modified too at the same time.
